relayer/relays/parachain/types.go

```diff
@@ -1,43 +1,131 @@
 package parachain
 
 import (
+	"fmt"
+	"math/bits"
+
 	"github.com/snowfork/go-substrate-rpc-client/v4/types"
 	"github.com/snowfork/snowbridge/relayer/chain/relaychain"
 	"github.com/snowfork/snowbridge/relayer/crypto/merkle"
 )
 
 // A Task is a bundle of items needed to submit commitments to Ethereum
 type Task struct {
-	ParaID                  uint32
-	BlockNumber             uint64
-	Header                  *types.Header
-	Commitments             map[ChannelID][]Commitment
-	BasicChannelBundleProof *MerkleProof
-	ProofInput              *ProofInput
-	ProofOutput             *ProofOutput
+	ParaID                        uint32
+	BlockNumber                   uint64
+	Header                        *types.Header
+	BasicChannelProofs            *[]MerkleProof
+	IncentivizedChannelCommitment *IncentivizedChannelCommitment
+	ProofInput                    *ProofInput
+	ProofOutput                   *ProofOutput
 }
 
 // A Commitment is data provably attested to by polkadot. The commitment hash
 // is contained in a parachain header. Polkadot validator nodes attest that the header
 // is genuine.
-type Commitment struct {
+type IncentivizedChannelCommitment struct {
 	Hash types.H256
-	Data interface{}
+	Data IncentivizedOutboundChannelMessageBundle
 }
 
-func NewCommitment(hash types.H256, data interface{}) Commitment {
-	return Commitment{
+func NewIncentivizedChannelCommitment(hash types.H256, data IncentivizedOutboundChannelMessageBundle) IncentivizedChannelCommitment {
+	return IncentivizedChannelCommitment{
 		Hash: hash,
 		Data: data,
 	}
 }
 
-type MerkleProof struct {
+type RawMerkleProof struct {
 	Root           types.H256
 	Proof          []types.H256
 	NumberOfLeaves uint64
 	LeafIndex      uint64
-	Leaf           []byte
+	// TODO: test that this decodes properly
+	Leaf []byte
+	// Leaf BasicOutboundChannelMessageBundle
+}
+
+type MerkleProof struct {
+	Root      types.H256
+	Proof     [][32]byte
+	HashSides []bool
+	Leaf      BasicOutboundChannelMessageBundle
+}
+
+func NewMerkleProof(rawProof RawMerkleProof, bundle BasicOutboundChannelMessageBundle) (MerkleProof, error) {
+	var proof MerkleProof
+
+	byteArrayProof := make([][32]byte, len(rawProof.Proof))
+	for i := 0; i < len(rawProof.Proof); i++ {
+		byteArrayProof[i] = ([32]byte)(rawProof.Proof[i])
+	}
+
+	hashSides, err := generateHashSides(rawProof)
+	if err != nil {
+		return proof, err
+	}
+
+	proof = MerkleProof{
+		Root:      rawProof.Root,
+		Proof:     byteArrayProof,
+		HashSides: hashSides,
+		Leaf:      bundle,
+	}
+
+	return proof, nil
+}
+
+func generateHashSides(commitmentProof RawMerkleProof) ([]bool, error) {
```

Comment:
Can we give this function a more generic signature. Would make it easier to test.

```suggestion
func generateHashSides(nodePosition int, breadth int) ([]bool, error) {
```

Reply:
I'd also add a few test cases, mostly to guard against future regressions.